src/kernel/pmm.zig

```diff
@@ -0,0 +1,175 @@
+
+    // Occupy the regions of memory that the memory map describes as reserved
+    for (mem.mem_map) |entry| {
+        if (entry.@"type" != MEMORY_AVAILABLE) {
+            var addr = std.mem.alignBackward(@intCast(usize, entry.addr), BLOCK_SIZE);
+            var end = @intCast(usize, entry.addr + (entry.len - 1));
+            // If the end address can be aligned without overflowing then align it
+            if (end <= std.math.maxInt(usize) - BLOCK_SIZE)
+                end = std.mem.alignForward(end, BLOCK_SIZE);
+            while (addr < end) : (addr += BLOCK_SIZE) {
+                setAddr(addr) catch |e| switch (e) {
+                    // We can ignore out of bounds errors as the memory won't be available anyway
+                    PmmBitmap.Error.OutOfBounds => break,
+                    else => panic(@errorReturnTrace(), "Failed setting address 0x{x} from memory map as occupied: {}", addr, e),
+                };
+            }
+        }
+    }
+    // Occupy kernel memory
+    var addr = std.mem.alignBackward(@ptrToInt(mem.physaddr_start), BLOCK_SIZE);
+    while (addr < @ptrToInt(mem.physaddr_end)) : (addr += BLOCK_SIZE) {
+        setAddr(addr) catch |e| switch (e) {
+            PmmBitmap.Error.OutOfBounds => break,
+            else => panic(@errorReturnTrace(), "Failed setting kernel code address 0x{x} as occupied: {}", addr, e),
+        };
+    }
+    log.logInfo("Done\n");
+    if (build_options.rt_test) {
+        runtimeTests(mem);
+    }
+}
+
+///
+/// Find the next free memory block, set it as occupied and return it. The region allocated will be of size BLOCK_SIZE.
+///
+/// Return: The address that was allocated.
+///
+pub fn alloc() ?usize {
+    if (bitmap.setFirstFree()) |entry| {
+        return entry * BLOCK_SIZE;
+    }
+    return null;
+}
+
+///
+/// Set the address as free so it can be allocated in the future. This will free a block of size BLOCK_SIZE.
+///
+/// Arguments:
+///     IN addr: usize - The previously allocated address to free. Will be aligned down to the nearest multiple of BLOCK_SIZE.
+///
+/// Error: Error || PmmBitmap.Error.
+///     Error.NotAllocated: The address wasn't allocated.
+///     PmmBitmap.Error.OutOfBounds: The address given was out of bounds.
+///
+pub fn free(addr: usize) (PmmBitmap.Error || Error)!void {
+    const idx = @intCast(u32, addr / BLOCK_SIZE);
+    if (try bitmap.isSet(idx)) {
+        try bitmap.clearEntry(idx);
+    } else {
+        return Error.NotAllocated;
+    }
+}
+
+pub fn runtimeTests(mem: *const MemProfile) void {
+    // Make sure that occupied memory can't be allocated
+    var prev_alloc: usize = std.math.maxInt(usize);
+    while (alloc()) |alloced| {
+        if (prev_alloc == alloced) {
+            panic(null, "PMM allocated the same address twice: 0x{x}", alloced);
+        }
+        prev_alloc = alloced;
+        for (mem.mem_map) |entry| {
+            if (entry.@"type" != MEMORY_AVAILABLE) {
```

Comment:
If possible, add a case to ensure you can't allocate memory within the kernel space

Reply:
I'm assuming you're talking about the kernel code space, which should be allocated anyway, so is the check necessary?

Reply:
The check is always necessary, it's a unit test, everything must be tested ;)

Reply:
Indeed, from the snippet I thought that this was in the `init` function :) I'll add a check.

Reply:
Isn't the kernel code space covered by the check underneath though? https://github.com/SamTebbs33/pluto/blob/b74addacf932b52199aae519dba99e5d0e600a29/src/kernel/pmm.zig#L143

Reply:
It is